tests/WP_SQLite_Driver_Metadata_Tests.php

```
@@ -0,0 +1,511 @@
+			$message  = sprintf(
+				'Error occurred while creating tables or indexes...<br />Query was: %s<br />',
+				var_export( $query, true )
+			);
+			$message .= sprintf( 'Error message is: %s', $err_data[2] );
+			wp_die( $message, 'Database Error!' );
+		}
+	}
+
+	public function testCountTables() {
+		$this->assertQuery( 'CREATE TABLE t1 (id INT)' );
+		$this->assertQuery( 'CREATE TABLE t2 (id INT)' );
+
+		$result = $this->assertQuery( "SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = 'wp'" );
+		$this->assertEquals( array( (object) array( 'COUNT ( * )' => '2' ) ), $result );
+
+		$result = $this->assertQuery( "SELECT COUNT(*) FROM information_schema.tables WHERE table_schema = 'other'" );
+		$this->assertEquals( array( (object) array( 'COUNT ( * )' => '0' ) ), $result );
+
+		// @TODO: The result key should be "COUNT(*)" instead of "COUNT ( * )".
+		//        The spacing was probably inserted by the translator.
+	}
+
+	public function testInformationSchemaTables() {
+		$this->assertQuery(
+			'
+			CREATE TABLE t (id INT PRIMARY KEY, name TEXT, age INT)
+		'
+		);
+
+		$result = $this->assertQuery( "SELECT * FROM information_schema.tables WHERE TABLE_NAME = 't'" );
+		$this->assertCount( 1, $result );
+		$this->assertRegExp( '/\d\d\d\d-\d\d-\d\d \d\d:\d\d:\d\d/', $result[0]->CREATE_TIME );
+		$this->assertEquals(
+			array(
+				'TABLE_CATALOG'   => 'def',
+				'TABLE_SCHEMA'    => 'wp',
+				'TABLE_NAME'      => 't',
+				'TABLE_TYPE'      => 'BASE TABLE',
+				'ENGINE'          => 'InnoDB',
+				'VERSION'         => '10',
+				'ROW_FORMAT'      => 'Dynamic',
+				'TABLE_ROWS'      => '0',
+				'AVG_ROW_LENGTH'  => '0',
+				'DATA_LENGTH'     => '0',
+				'MAX_DATA_LENGTH' => '0',
+				'INDEX_LENGTH'    => '0',
+				'DATA_FREE'       => '0',
+				'AUTO_INCREMENT'  => null,
+				'CREATE_TIME'     => $result[0]->CREATE_TIME,
+				'UPDATE_TIME'     => null,
+				'CHECK_TIME'      => null,
+				'TABLE_COLLATION' => 'utf8mb4_general_ci',
```

Do we support different collations? Or is it always the same underlying collation?

---

I wonder if we should throw an error on collation mismatch, or at least a warning. I guess some apps will try to use collation that’s not supported in sqlite, but it won’t really matter in at least some of them. A warning perhaps? Or a wp amin notice?

---

@adamziel At the moment, we don't apply collations in any way, so internally, our encoding is `utf8mb4` and the collation probably something like `utf8mb4_unicode_ci`, `utf8mb4_general_ci`, or `utf8mb4_0900_ai_ci` (I need to check what exactly happens in SQLite and which collations they support).

> I wonder if we should throw an error on collation mismatch, or at least a warning.

Good question. I think it's tricky to be strict here because even though WP Core embraced `utf8mb4`, some plugins may still have not done that properly, even though they could be saving Unicode bytes in some columns. It seems to me like this may need a bit more investigation. Maybe I should create a ticket?

---

Yes, let's create a ticket. As for the next steps, I'd suggest:

* Throw an exception if we see a collation referring to a different character set or comparison rules
* Add an option like `enforce_utf8_charset` that defaults to false for all the other cases. When it's `true`, we'd log a warning and use a sensible default encoding when we see a query with a "salvageable" encoding or collation definition. When it's `false`, we'd throw a fatal error explaining what happened and that there's an option you can use. It could also be called `strict_mode` or so.

What's "salvageable" is quite arbitrary and it's easier to say what isn't.

For example, an incompatible set of collation rule would lead to a very different application behavior and I'd just throw an error right away. [MySQL Collation doc page](https://dev.mysql.com/doc/refman/8.0/en/charset-collation-names.html) explains different parts of the collation suffix:

```
Suffix | Meaning
-- | --
_ai | Accent-insensitive
_as | Accent-sensitive
_ci | Case-insensitive
_cs | Case-sensitive
_ks | Kana-sensitive
_bin | Binary
```

When a table declares `latin1`, `utf16`, or anything that isn't `utf8`, we'd likely break the app by quietly using utf-8.

On the flip side, I think we're good to rewrite `utf8`, `utf8mb3`, and other similar variations as `utf8mb4` (when the option is set). [Unicode characters sets](https://dev.mysql.com/doc/refman/8.4/en/charset-unicode-sets.html) page lists deprecated character sets and recommends using `utf8mb4` instead. That would still change the application behavior, but I can't imagine a plugin that relies on collating up to 3 bytes from every UTF-8 characters and not the fourth byte. Similarly, `utf8mb4_general_ci` is deprecated in favor of `utf8mb4_unicode_ci` and I think we could treat them as the same charset.

The [Unicode character sets](https://dev.mysql.com/doc/refman/8.4/en/charset-unicode-sets.html) page also discusses other variations, such as `general_`, language-specific character sets, etc. It's important that we're aware of this general problem space and, when in doubt, default to throwing an error instead of continuing silently.

---

I don't mean it as a blocker for this PR, by the way. We'd just need to add, at a very least, the exception throwing in the very next PR to avoid creating an illusion of a working application.

---

@adamziel I created the ticket: https://github.com/Automattic/sqlite-database-integration/issues/25

I'm curious which of the collations and suffixes we'll be able to correctly support in SQLite. SQLite has some collation support, but I haven't explored that yet.

Once we start throwing the exception, I'm also curious if it will cause any Playground Tester test cases to fail.